kinematics_interface_kdl/src/kinematics_interface_kdl.cpp

```diff
@@ -20,232 +20,229 @@ namespace kinematics_interface_kdl {
-    double tmp[] = {frame_.p.x(), frame_.p.y(), frame_.p.z()};
-    // KDL::Rotation stores data in row-major format. e.g Xx, Yx, Zx, Xy... = data index at 0, 1, 2, 3, 4...
-    for (int r = 0; r < 3; r++) {
-      for (int c = 0; c < 3; c++) {
-        transform_vec[r + 4 * c] = frame_.M.data[3 * r + c];
-      }
-      transform_vec[4 * 3 + r] = tmp[r];
-    }
-    transform_vec[15] = 1.0;
+  // special case: since the root is not in the robot tree, need to return identity transform
+  if (link_name == root_name_)
+  {
     return true;
   }
 
-  bool KDLKinematics::update_joint_array(const std::vector<double> &joint_pos) {
-    memcpy(q_.data.data(), joint_pos.data(), joint_pos.size() * sizeof(double));
-    return true;
+  // create forward kinematics solver
+  fk_pos_solver_->JntToCart(q_, frame_, link_name_map_[link_name]);
+  double tmp[] = {frame_.p.x(), frame_.p.y(), frame_.p.z()};
+  // KDL::Rotation stores data in row-major format. e.g Xx, Yx, Zx, Xy... = data index at 0, 1, 2, 3, 4...
+  for (int r = 0; r < 3; r++)
+  {
+    for (int c = 0; c < 3; c++)
+    {
+      transform(r + 4 * c) = frame_.M.data[3 * r + c];
+    }
+    transform(4 * 3 + r) = tmp[r];
   }
+  return true;
+}
 
-  bool KDLKinematics::verify_link_name(const std::string &link_name) {
-    if (link_name == root_name_) {
-      return true;
-    }
-    if (link_name_map_.find(link_name) == link_name_map_.end()) {
-      std::string links;
-      for (auto i = 0u; i < chain_.getNrOfSegments(); i++) {
-        links += "\n" + chain_.getSegment(i).getName();
-      }
-      RCLCPP_ERROR(LOGGER, "The link %s was not found in the robot chain. Available links are: %s",
-                   link_name.c_str(), links.c_str());
-      return false;
-    }
-    return true;
-  }
+bool KDLKinematics::update_joint_array(const Eigen::Matrix<double, Eigen::Dynamic, 1> &joint_pos)
+{
+  memcpy(q_.data.data(), joint_pos.data(), joint_pos.size() * sizeof(double));
+  return true;
+}
 
-  bool KDLKinematics::verify_transform_vector(const std::vector<double> &transform) {
-    if (transform.size() != 16) {
-      RCLCPP_ERROR(LOGGER,
-                   "Invalid size (%zu) for the transformation vector. Expected size is 16.",
-                   transform.size());
-      return false;
-    }
+bool KDLKinematics::verify_link_name(const std::string & link_name)
+{
+  if (link_name == root_name_)
+  {
     return true;
   }
-
-  bool KDLKinematics::verify_cartesian_vector(const std::vector<double> &cartesian_vector) {
-    if (cartesian_vector.size() != jacobian_->rows()) {
-      RCLCPP_ERROR(LOGGER,
-                   "Invalid size (%zu) for the cartesian vector. Expected size is %d.",
-                   cartesian_vector.size(), jacobian_->rows());
-      return false;
-    }
-    return true;
+  if (link_name_map_.find(link_name) == link_name_map_.end())
+  {
+    std::string links;
+    for (auto i = 0u; i < chain_.getNrOfSegments(); i++)
```

Comment:
```suggestion
    for (auto i = 0u; i < chain_.getNrOfSegments(); ++i)
```

++i is slightly more performant. Read why here:  https://betterprogramming.pub/stop-using-i-in-your-loops-1f906520d548